nalgebra-sparse/src/cs.rs

```diff
@@ -543,3 +544,146 @@ pub fn convert_counts_to_offsets(counts: &mut [usize]) {
+        ));
+    }
+
+    // Check that the first and last offsets conform to the specification
+    {
+        let first_offset_ok = *major_offsets.first().unwrap() == 0;
+        let last_offset_ok = *major_offsets.last().unwrap() == minor_indices.len();
+        if !first_offset_ok || !last_offset_ok {
+            return Err(SparseFormatError::from_kind_and_msg(
+                SparseFormatErrorKind::InvalidStructure,
+                "First or last offset is incompatible with format.",
+            ));
+        }
+    }
+
+    let mut minor_index_permutation: Vec<usize> = Vec::new();
+
+    // Test that each lane has strictly monotonically increasing minor indices, i.e.
+    // minor indices within a lane are sorted, unique. Sort minor indices within a lane if needed.
+    // In addition, each minor index must be in bounds with respect to the minor dimension.
+    {
+        for lane_idx in 0..major_dim {
+            let range_start = major_offsets[lane_idx];
+            let range_end = major_offsets[lane_idx + 1];
+
+            // Test that major offsets are monotonically increasing
+            if range_start > range_end {
+                return Err(SparseFormatError::from_kind_and_msg(
+                    SparseFormatErrorKind::InvalidStructure,
+                    "Offsets are not monotonically increasing.",
+                ));
+            }
+
+            let minor_idx_in_lane = minor_indices.get(range_start..range_end).ok_or(
+                SparseFormatError::from_kind_and_msg(
+                    SparseFormatErrorKind::IndexOutOfBounds,
+                    "A major offset is out of bounds.",
+                ),
+            )?;
+
+            // We test for in-bounds, uniqueness and monotonicity at the same time
+            // to ensure that we only visit each minor index once
+            let mut iter = minor_idx_in_lane.iter();
+            let mut prev = None;
+            let mut nonmonotonic = false;
+
+            while let Some(next) = iter.next().copied() {
+                if next >= minor_dim {
+                    return Err(SparseFormatError::from_kind_and_msg(
+                        SparseFormatErrorKind::IndexOutOfBounds,
+                        "A minor index is out of bounds.",
+                    ));
+                }
+
+                if let Some(prev) = prev {
+                    if prev == next {
+                        return Err(SparseFormatError::from_kind_and_msg(
+                            SparseFormatErrorKind::DuplicateEntry,
+                            "Input data contains duplicate entries.",
+                        ));
+                    }
+                    if prev > next {
+                        nonmonotonic = true;
+                    }
+                }
+                prev = Some(next);
+            }
+
+            // sort if indices are nonmonotonic and sorting is expected
+            if nonmonotonic && sort {
```

Comment:
What if the indices are `nonmonotonic` but we're *not* asked to sort? In this case we need to return failure, right?

Reply:
It doesn't look to me like this was addressed yet?

Reply:
It should be addressed after adding following lines of code: 
https://github.com/dimforge/nalgebra/pull/1015/commits/9ff5d1a97929d4a9494383eb6dc5590f5243491f#diff-ad9fa9a189df6658e5fee89093cd2efba6292df75401887d4c3e336a6fca2e4bR650

Reply:
Ah, right, I missed that. Do the tests currently exercise this branch?